compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -253,9 +254,14 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
 
     // turn a branch with all successors identical to a goto
     fn simplify_branch(&mut self, terminator: &mut Terminator<'tcx>) -> bool {
-        match terminator.kind {
-            TerminatorKind::SwitchInt { .. } => {}
-            _ => return false,
+        // Removing a `SwitchInt` terminator may remove reads that result in UB,
+        // so we must not apply this optimization when mir-opt-level = 0.
+        if self.mir_opt_level == 0 {
```

Comment:
We could perhaps optimize trivial operands... Like plain locals. Maybe consts?

Reply:
Well I guess even plain locals could have UB if they're uninitialized. But Rust would never produce that 🤔

Reply:
Ah, not just a bug for Miri then but a bigger problem. Certainly we shouldn't optimize before borrowck.

So instead of checking the mir_opt_level, this should be told which "phase" it is in (or it can maybe even get that info from the MIR body, not sure), and then not do this transformation in the "analysis" phase.